awx/api/serializers.py

```diff
@@ -3286,18 +3276,12 @@ def validate(self, attrs):
 
         extra_vars = attrs.get('extra_vars', {})
 
-        if isinstance(extra_vars, basestring):
-            try:
-                extra_vars = json.loads(extra_vars)
-            except (ValueError, TypeError):
-                try:
-                    extra_vars = yaml.safe_load(extra_vars)
-                    assert isinstance(extra_vars, dict)
-                except (yaml.YAMLError, TypeError, AttributeError, AssertionError):
-                    errors['extra_vars'] = _('Must be valid JSON or YAML.')
-
-        if not isinstance(extra_vars, dict):
-            extra_vars = {}
+        try:
+            extra_vars = parse_yaml_or_json(extra_vars, silent_failure=False)
+        except:
```

Comment:
In the big picture (and I'm willing to do this now if it's highly requested), I would like to have `parse_yaml_or_json` take all the anticipated exception cases (ValueError, TypeError, etc.) and convert it into a new custom type error. To do that right, however, there are several properties I think we should set on the exception class, and a lot to do in order to have a highly descriptive level of verbosity in the error message.

After we do that, we can catch _just_ the custom exception type, and then if an unexpected error comes up in the parser, it will surface as a normal 500. This will also allow us to give vastly more verbosity than "Must be valid JSON or YAML". We could say "JSON parser returned a non-dict type, you must provide a dict", and you could see this getting very specific to the particular code path that it failed in.

Reply:
Yes, that's exactly what I believe the net effect for lines of code such as this in the serializer should be. Good suggestion.

My grand hope is that the ParseError exception could also return an incredibly specific error message, like `{"extra_vars": "Failed to parse as JSON due to 'TypeError in position xxx in line xxx', and YAML parser returned a non-dictionary."}`, with some additional error handling in the utility method.